File: src/java/3_StructuralPatterns/1_Adapter/AdapterExercise.java
package com.activemesa.structural.adapter.exercise;

class Square
{
  public int side;

  public Square(int side)
  {
    this.side = side;
  }
}

interface Rectangle
{
  int getWidth();
  int getHeight();

  default int getArea()
  {
    return getWidth() * getHeight();
  }
}

class SquareToRectangleAdapter implements Rectangle
{
  private Square square;

  public SquareToRectangleAdapter(Square square)
  {
    this.square = square;
  }

  @Override
  public int getWidth()
  {
    return square.side;
  }

  @Override
  public int getHeight()
  {
    return square.side;
  }
}

File: src/js/3_StructuralPatterns/1_Adapter/adapter_exercise.js
class Square
{
  constructor(side)
  {
    this.side = side;
  }
}

function area(rectangle)
{
  return rectangle.width * rectangle.height;
}

class SquareToRectangleAdapter
{
  constructor(square)
  {
    this.square = square;
  }

  get width() {
    return this.square.side;
  }

  get height() {
    return this.square.side;
  }
}

describe('adapter', function()
{
  it('should adapt things, duh!', function()
  {
    let sq = new Square(11);
    let adapter = new SquareToRectangleAdapter(sq);
    expect(area(adapter)).toEqual(121);
  });
});